### ISSUE TITLE
Investigation: Generate mipmaps

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#386]
Generating mipmaps is something that most 3D applications want to do. As textures are often the largest asset 3D apps need to download, generating the smaller miplevels can save significant downloading time, causing the app to be able to launch earlier. We've heard that download times are a significant concern to Web game developers; there's a strong inverse correlation between how long a games takes to load and how many users actually end up playing the game.

WebGPU currently provides no facilities to generate mipmaps. It is true that a 3rd party library could do this, but taking an additional dependency for something that most 3D authors will want to do seems like bad design. Indeed; the purpose of doing this in the browser is to decrease download sizes, and adding an additional framework dependency is contrary to that goal.

WebGL solved this by including [WebGLRenderingContext.generateMipmap()](https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.8). Developers clearly desire using this; a quick [search](https://github.com/search?l=JavaScript&q=generateMipmap&type=Code) of GitHub shows that it's used more than 46,000 times. Indeed, at least one developer is using this WebGL API to generate mipmaps and then passing the results into WebGPU.

WebGPU should provide built-in facilities for generating mipmaps. Of course, we shouldn't force developers to use the built-in facilities. It is true that different developers desire mipmap filtering algorithms, and they should be free to write their own mipmap generation code. However, for most authors, the presence of built-in mipmap generation facilities will be the difference between mipmaps existing in their app at all, and mipmaps not existing in their app.

## Metal
It's a [method](https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400748-generatemipmapsfortexture?language=objc) on `MTLBlitCommandEncoder`. It is executed on the GPU, so the command needs to be submitted to the queue and is therefore ordered with relation to other GPU operations. Also, because this is inside a command encoder, all compute/graphics command encoders need to be closed before this can be issued.

The docs have a note:
> The filtering used to generate the mipmaps is implementation-dependent and may vary by Metal feature set.

## Direct3D 12
The core API doesn't have any support for generating mipmaps. However, the Microsoft-authored DirectX Tool Kit 12 [includes support](https://github.com/microsoft/DirectXTK12/wiki/ResourceUploadBatch) for it. The docs say that it's implemented on top of a compute shader.

> The GenerateMips function uses a DirectCompute shader to perform the mip-generation using bi-linear interpolation

Direct3D 12 doesn't have passes, so there are no concerns there. However, the API isn't a one-shot generation; instead, it's a `ResourceUploadBatch` object that has to be opened and closed.

## Vulkan
Vulkan doesn't have any built-in facilities for generating mipmaps (because of course it doesn't). It [looks like](https://vulkan-tutorial.com/Generating_Mipmaps) you can do it with (a ton of code around)`vkCmdBlitImage`, which means that any render passes need to be closed.

## Compute shader
Even if we didn't want to use the above facilities, we could still implement it using a compute shader. This would require that all render passes are closed.

## Recommendation
In keeping with the current design of putting blit-style commands directly on the `GPUCommandEncoder`, we can add a `generateMipmaps(texture)` call there. This will allow for us to require that any render or compute passes are closed before it's called. Such a design should be compatible with each of the above approaches.

**Comment by kvark** [comment]
You've almost lost me at the first sentence :)
> Generating mipmaps is something that most 3D applications want to do.

Most 3D applications that are performance-sensitive wouldn't want to generate mipmaps at run-time, except for some cases where texture content is not known in advance (procedurally generated, captured from the environment, etc).

> As textures are often the largest asset 3D apps need to download, generating the smaller miplevels can save significant downloading time, causing the app to be able to launch earlier.

Mipmaps constitute for 1/3 of the original texture volume. Besides, an application can start perfectly fine while the mipmap levels are still being transferred. The best way for an app to start quickly is to pre-generate the mipmaps and start loading the higher (smaller) mips first, which would allow it to start much faster than waiting for the full textures. Once other mipmap levels arrive, an application can issue `copyBufferToTexture` instructions to put it into the texture. Obviously, this option is very different from the approach suggested in this PR.

> It is true that a 3rd party library could do this, but taking an additional dependency and sending a library across the wire for something that most 3D authors will want to do seems like bad design.

Is having a 3rd party dependency such a big trouble in JS/WASM world? I'm surprised we'd be trying to reduce the ecosystem as opposed to relying on it to grow.

Previously with `setSubData` we concluded that the users can just copy-paste a small function implementing this into their program. Is the main difference here the fact that a shader is involved? If that's the case, we may consider just providing `blitImage` method in our render pipelines:
  - it's fairly non-controversial, has been around since GL days
  - can be used to fill out the mip maps in a few lines of code

**Comment by magcius** [comment]
* Mipmaps need to be filtered differently depending on whether they are used as wrap or clamp (you would want to filter edge pixels differently)
* Compressed textures cannot have mipmaps generated at runtime, unless we require shipping a BCn compressor
* For any sort of complex PBR model, you want to do special things to your mipmaps anyway (convolve normals with roughness, prefilter env maps)

I would encourage that if the WebGPU WG would like runtime mipmap compression, that this is shipped as a library developed inside the WG and released as open-source for application authors to use.

**Comment by prideout** [comment]
Including a simple blitting function instead of a mipmap generator would be less controversial.  Note that Vulkan provides `vkCmdBlitImage` with somewhat configurable filtering, but fixed to CLAMP_TO_EDGE wrapping behavior.

**Comment by magcius** [comment]
I agree with including a blit image command, though I wouldn't encourage using it in place of offline mipmap generation if possible.

**Comment by deltakosh** [comment]
You need to see the big picture here. We have thousands of users who are relying on us (babylon.js for instance but pretty Three has the same concern) to generate their mipmaps. Not every developer is a AAA game developer.

**Comment by kvark** [comment]
@deltakosh interesting that you mention this, because Three.js and Babylon.js are in a good position to implement it once and for all in their own codebase, using the most efficient methods available.

**Comment by deltakosh** [comment]
So far (for webGPU) we are using webgl so I'm not sure it is the most efficient method :)
I do not care too much if this is a separate library but we try to reduce external dependencies as much as possible so I would prefer having it directly available in the browser

**Comment by kvark** [comment]
@deltakosh is your use of WebGL justified by the inability to generate mipmaps with WebGPU directly? I mean, just having a shader that samples an image, and then running it for each level is not too difficult for you as a framework author.

**Comment by deltakosh** [comment]
It is not, I agree and as I stated earlier we will do it if no other option is available. We are already resampling textures using shaders to make them POT so no big deal. But again it is more about the overall usability of the API.

**Comment by sebavan** [comment]
I am wondering why we try to push a lot of the functionalities to external lib and outside of the browser api scope. I am thinking the browser APIs is an amazing place to simplify the user code and prevent everybody from writing again the same kind of code.

Making it a lib means we expect people to use it so why not directly integrating in the browser ? (Even more as it is not just a trivial task to do well).

I am afraid that currently to get started I would need to install a lib for mipmaps, one to deal with buffers upload, maybe a compiler, and probably some other utils. 

What is the main issue of keeping some default utils in the spec ? Like uploading imageData, we could argue that we do not need it but is it not good to have convenience method like those for the devs ?

**Comment by magcius** [comment]
I would rather not have browser-implemented helper libraries, unless they're all guaranteed to be identical (and this is a *very* tricky thing to ensure). There isn't a one-size-fits-all solution, and there are a lot of constraints. The blit image approach won't work for filtering cubemap images, for instance.

We're already seeing a huge delay getting the WebGPU MVP added into browsers. If we add it into the spec, it should not be added for MVP, and Babylon will have to ship mipmap generation code anyway...

**Comment by KeepItOneHundred** [comment]
IMO the problem isn't so much with having built-in helpers, it's those helpers incentivizing developers to write inefficient code. If the browser can generate mipmaps on-the-fly then the path of least resistance is to use uncompressed textures and waste a ton of memory/bandwidth, as was the case with WebGL.

I'd rather see the effort put towards supporting Basis Universal as a first-class feature with transcoding provided by the browser, so the easy path is also the fast path.

**Comment by deltakosh** [comment]
Well if this is not done by the browser it will be done by the frameworks so this will not incentive users much to pick Basis path

**Comment by sebavan** [comment]
I would say that even with more compressed format and so on, we still need sometimes for procedural textures, textures from video or webcam capture, or even texture created from another canvas some easy path to mip maps.

I totally get the MVP timing requirements and I do not think there are any rush for helpers in the spec. In frameworks like babylon or three we will be able to go around it but as a general API surface on a longer term than the MVP, i think it still makes sense to have methods simplifying the devs life.

I might misunderstand the overall audience, but if we are not only targeting big engines out there, and we want to expand our reach, having convenience helpers is probably a good way forward. 

Also as in browsers, the less call to the native APIs, the faster, it might be a tiny performance boost compared to doing all of the calls manually. It might even help reducing the JS file size which more and more people seem to take as a strong concern and metric.

**Comment by jdashg** [comment]
I strongly believe this should be handled by applications and libraries.

WebGL's GenerateMipmaps is unstandardized(!), poor (unguaranteed!) quality, and has a number of restrictions that surprise users (filterable+renderable formats only!). While superficially useful, it's pretty limited, and I don't think is something we should try to replicate.

**Comment by RafaelCintron** [comment]
@litherum write:
> Direct3D 12 doesn't have passes, so there are no concerns there

The latest versions of D3D12 have [render passes](https://docs.microsoft.com/en-us/windows/win32/direct3d12/direct3d-12-render-passes).

**Comment by krogovin** [comment]
One important use case of mipmap generation is when a 3D scene renders to an offscreen buffer which is then textured directly in way that the sampling is not uniform.

From the point of view of hardware and drivers, many GL drivers implement a highly tuned implementation of mipmap generation potentially closely tied to the GPU architecture, so those might win performance over one made from a library. I can imagine a scenario where a tiled based GPU could make it faster if the driver new as part of the render target that mipmaps where needed, but there is no API doing that so I doubt any GPU does that.

On the otherhand, most of the time driver-done mipmap generation is a simple box-filter which is naturally the cheapest but not necessarily the best looking one.

**Comment by kainino0x** [comment]
AFAICT, neither Vulkan nor D3D12 implements a GenerateMipmaps, so we don't have access to any highly optimized driver implementation, so the browsers would have to write our own.

**Comment by grorg** [comment]
Discussed at the [2019-08-12 meeting](https://docs.google.com/document/d/18Ze0h9InJN5QWXe0iN9UzEebfiEIg8HYo-aX9rKvhuk/edit).

**Comment by litherum** [comment]
> AFAICT, neither Vulkan nor D3D12 implements a GenerateMipmaps ... so the browsers would have to write our own

See [above](https://github.com/gpuweb/gpuweb/issues/386#issue-476292952). The Microsoft-authored DirectX Tool Kit 12 [includes support](https://github.com/microsoft/DirectXTK12/wiki/ResourceUploadBatch) for it.

**Comment by litherum** [comment]
> While superficially useful

It's hard to argue that 46,000 usages on GitHub is only superficially useful.

**Comment by litherum** [comment]
> a library developed inside the WG

I think we'd have to modify our charter to do this.

**Comment by kvark** [comment]
> It's hard to argue that 46,000 usages on GitHub is only superficially useful.

To be fair, WebGL developers use it because OpenGL has it. We don't have this kind of legacy with WebGPU. So, I wouldn't treat this number as a strong indication that the built-in support is needed.

**Comment by litherum** [comment]
If you consider OpenGL usages of this function, the [count](https://github.com/search?l=C&q=glGenerateMipmap&type=Code) grows to over 200,000 usages. The fact that it's used so often in real code means these facilities are more than superficially useful.

**Comment by devshgraphicsprogramming** [comment]
If you have the time and effort to do it, do it, just make sure that:
1. Your Mip-Map Generation runs in compute shader
2. That Compute Shader can do 4 or more mip-map levels in a single dispatch
3. The mip-mapping within a dispatch happens in shared memory
4. You can batch multiple textures to be mip-mapped at once
5. You can control the filter used and provide at least the Kalman filter

If you provide any less than this then its simply not worth the effort.

@litherum you're right that everyone relies on OpenGL's and WebGL's mip-map generation, however they also accept (what the WebGPU WG has always been allergic to) *inconsistency*.

The OpenGL (and WebGL) spec leave it completely up to the implementation as to:
- The Filter Used (mostly the frequency destroying box filter)
- The filtering of non PoT mip-maps
- The Speed and Efficiency
- The Result (different mip-maps of the same base texture can look vastly different on 2 implementations)

Also mip-mapping a block compressed texture is a major pain.

**Comment by devshgraphicsprogramming** [comment]
The only way to provide this feature is the way @magcius suggested, otherwise textures with minification will look different on different browsers (and possibly even backends/GPUs).

**Comment by kvark** [comment]
I developed a benchmark suite for comparing Vulkan blits versus transfers and shader draws, see https://github.com/gfx-rs/gfx/pull/2960#issuecomment-522157401
TL;DR: on NVidia GPU there isn't a convincing difference.

**Comment by hjlld** [comment]
This spec is concerned about GPU but firstly it is a Web spec.

On native few AAA games use mipmap in runtime, because they don't need to exchange textures with users, and from the perspective of performance, pre-generating mipmaps and using compressed texture is obviously the best choice.

But on Web, many apps are created for interactiving with users, besides capture image from web cameras, there're many 3D configurators based on WebGL in use in productive environment right now which need user to upload a bitmap or draw something on canvas as a texture. Web developers and designers do not have so many options to choose like native developers. A quick, easy and even dirty way to realize some features to resolve user's pain is usually used in web development.

It's irresponsible leaving this to 3p libs, which will lead lots of other promblems like security( npm security stories ), maintenance( abandoned by authors ), and so on. 

Maybe web is more important than gpu in WebGPU.

**Comment by krogovin** [comment]
Needing to generate mipmaps at run time is going to be almost always from rendering to texture use cases and needing those textures' mipmaps.

**Comment by litherum** [comment]
Not just rendering to texture. In WebGPU just as in WebGL, there are many sources of textures such as video frames, 2d canvases, HTMLImageElements, etc, which would usually need mipmaps when being rendered in a 3D environment.

**Comment by toji** [comment]
Late to the conversation, but I figured I'd add my recent experience on the matter. I was doing some WebGPU investigations in Chrome Canary and implemented my own mipmap generation function as part of it, which [you can see here](https://github.com/toji/webgpu-test/blob/master/js/webgpu-renderer/webgpu-texture-helper.js) if you're so inclined. It had to work around a couple of Chrome-specific bugs, and I didn't delve into compute shaders to do it, so it's not optimal by any stretch of the imagination, but it seems representative of what many developers would build if left to their own devices. It left me with a few observations.

First, and most importantly, I want to stress that while it's absolutely the case that the optimal thing to do it to generate your mips before hand, and I fully expect support for BASIS and the like to improve the average developer's inclination to do that, it can't be understated how much the web ecosystem encourages uses that fall outside the optimal paths. It's very common for developers to use WebGL to render images that come from sources they don't fully control, including video elements, and it's not unreasonable for them to want those resources to look decent. This was stated above but it's worth repeating.

For my part, while building the mip generation code wasn't hard (if you ignore the Chrome bugs) I did find myself wishing I had something equivalent to the aforementioned vkCmdBlitImage, which would reduce the entire operation from a full shader/pipeline/render pass creation to a simple loop of blit commands from one level to the next. It feels like providing a utility function along those lines would not only help greatly simplify this particular common task but also make a variety of other use cases more palatable as well. I recognize that there may be some concern about differences in how such a thing is implemented between native APIs/User Agents, but I'd argue that when developers are pulling in random images from Twitter/Reddit/etc for their apps they're unlikely to be very concerned about the minutia of the filtering algorithm used. For those that are, more advanced mip generation methods are always possible.

And last, a technical observation/inquiry: If you expect that doing mip generation the way I did, without a compute shader, will be common then that implies that every texture that has mipmap's generated for it will need to be created with the OUTPUT_ATTACHMENT usage. I don't have any sense of if that's detrimental or not, but it felt strange to me. Is this something to be avoided or does it really not matter much in the end?

**Comment by Kangz** [comment]
Thanks for the constructive feedback @toji. I made a [slight adaptation of your code](https://gist.github.com/Kangz/782d5f1ae502daf53910a13f55db2f83) to show what mipmap generation with repeated render-passes would look like without Chromium's issues (and also using `getBindGroupLayout`).

I am worried too about the cost of adding the `OUTPUT_ATTACHMENT` usage to texture that are only going to be used for sampling afterwards. Maybe the helper function could generate mipmaps in a staging texture then copy each mip to the real texture if the texture doesn't need to be renderable. A `vkCmdBlitImage`-like method would help but still have the same problem for `OUTPUT_ATTACHMENT`.

For developers just using WebGPU that need mipmaps but otherwise don't care about their quality, do you think that a helper library like the one in the Gist above would work if it were in a prominent enough place?

**Comment by toji** [comment]
I wasn't aware of `getBindGroupLayout`. That's a nice simplification! And yes, I think that having helper libraries like that available would help significantly, especially if it was hosted in some repository closely related to WebGPU (maybe even the spec repo itself.) If the code was compact enough you could even take it a step further and write example mimpap generation into the spec itself as non-normative example code.

Regardless, though, I think that having some simple, flexible helper methods as part of the API itself that do the "right thing" for 90% of developers needs is a positive thing when it comes to a web API, where it may not be for an equivalent native API. And in this case I mean something like a `blitTextureToTexture()` method rather than a full on `generateMipmaps()`. Feels like that would fit nicely with `bufferSubData()` style methods for simplifying common buffer uploads and the already existing `copyImageBitmapToTexture()` for simplifying texture uploads.

**Comment by litherum** [comment]
> Regardless, though, I think that having some simple, flexible helper methods as part of the API itself that do the "right thing" for 90% of developers

I agree 100%.

**Comment by deltakosh** [comment]
Can’t agree more

**Comment by jdashg** [comment]
Some sort of filtering image-to-image blit entrypoint is appealing to me, and I think makes generateMipmap in userland a loop away. IIRC @kvark mentioned there's prior art in our target APIs, and that some hardware (Intel?) could benefit quite a lot from using them, vs doing a renderpass to sample-and-draw or something.

**Comment by Nehon** [comment]
Hello, I think I'm going to chime in too, because I'm a bit worry, seeing some arguments in this thread.
I'm Rémy, I'm a 3D developer at Sketchfab. We are using webgl1 and webgl2 and we are closely following webGPU as we see it as the next step for us.

To give you some context, Texture data is by far what makes the most size of a sketchfab model when you download it. Individual texture size is something we monitor and optimize a lot. Pre generating mipmaps and sending them over the network is not a practical, and even a realistic solution. Seeing arguments like "it's only 1/3 of the texture data" is scary, to say the least. We are targeting any kind of hardware and any kind of network bandwidth. We can't rely in the fact that all our users will have a killing fiber connection. Most of them have a slow 3g bandwidth.

Transport of the data is what should matter first in a Web context.

We use glGenerateMipMap for our textures, and we use them in an uncompressed format on the GPU. 
The reason is, compressed formats on the GPU are just not practical in real life. Support is not reliable enough on different platforms and the cost of having a different code path when it's supported outweighs the benefit. Also in terms of transport size, jpeg, with the right compression parameters, is still the king.
Note that what's also important in a compression format, is the time it takes to compress. Users upload textures in jpg or png format by the thousandths every day on sketchfab, and we process them all. For example, we gave up on ASTC, for this reason. It's too slow to compress files.
In order to mipmap Framebuffer attachment textures (when we need to control the filtering) we do multiple render passes into a texture.

WebGPU success is tied with finding the right balance in ease of use, standardization and flexibility. I wouldn't make assumptions of what users will do, because "that's what they should do". Please, give users some flexibility to adapt to their real life use cases.

I recently fiddled into mipmaps generation and that's how I found this thread. So here are my 2 cents on how I'd like to see it in webGPU:
- Mipmap generation HAS to be possible at runtime.
- Having 1 hardware command to generate mipmaps on regular textures (as opposed to RTT textures) is REALLY handy. Setting up several passes for each textures is possible of course, but depending on the architecture it can be a pain. Also, I would see it as a regression of webGPU compared to webgl. I've seen arguments like "It was not reliable in webgl so let's not do it in webgpu". The reason it was not reliable is because it's poorly described in opengl specs an mostly left to the implementer interpretation, and this leads to fragmentation. The point of making a new API is to make things right, so just make it right by describing in the specs EXACTLY what this process should be. Everyone bundling its own mipmap implem in each application doesn't feels like a wise move for the big picture of the web in general to me.
- Having a way to specify the filtering kernel for mipmaps. This one would be nice to have. For RTT you may need special filtering for things like SSAO and so on. Being able to pass in a filtering kernel to generateMipMap would be super nice, but I understand it complexifies everything, and I wouldn't mind if it's not there, as long as the following point works. 
- Being able to read from a mip level and write into another WITHOUT throwing a feedback loop error. This is possible in native opengl or openglES3. It's not clear if it should be possible in webgl2, but in practice it fails (again, not blaming implementers, the specs are really unclear about it). In webgl1 we resort to put the mipmaps in a texture atlas, and in webgl2 we render the mipmap to another buffer and then copy it into the right mipmap level of the source texture... This is super cumbersome.. nobody wants to do this. We would like to be able to read from one mipmap level to render the next one using the next mipmap level as render target.
 
If you want the API to be frictionless, make all these plumbing operations as seamless as they can be.

------ TL;DR ------
In my humble opinion
1) Have a GenerateMipMap routine, thoroughly described in the spec, for the general use case is a must
2) Being able to render to a MipMap level while reading from the previous mipmap level, for cases where you need to control the filter kernel

If 1) is a no, then a khronos official generateMipMap routine build on top of the API would be very much appreciated.

**Comment by krogovin** [comment]
I will add some of my two cents to the above:

- The reason for not allowing feedback loops is mostly for tiled based renderers. This is why the handy `glTextureBarrier()` does not exist on mobile.
- However, the ability to read from one mipmap level and write to another should be allowed. The main issue is that the shader must never access the mipmap it is writing to. This is the exact use case for mipmap generation without compute. Sadly, the feedback loop checking only happened at the API level and never drilled down to the shaders to see if the feedback look was not present. This is a major hassle for me as well as the ping-ponging also costs extra blits which means extra bandwidth which is even more critical to save on mobile.

My thoughts are this: since the texture access stuff already has listed what mipmap level range is to be accessed, then there should not be an issue with feedback loop for mipmap generation via drawing triangles, i.e. if an image thingy is listed to only read form mipmap levels [0, N) and rendering is to level N, then that should never be considered a feedback loop.

However, none of this prevents using ImageStore in a fragment shader with an attachmentless framebuffer active. The main issue with using ImageStore is that usually, but not always, is that ImageStore can be slower on some hardware compared to letting the pixel backend do the (heavily cached) writes. Compute can be worse that a fragment shader because the most optimal group size can be quite GPU specific (though compute does have the benefit that helper pixels do not exist, so it won't spawn "double" on the diagonal).

**Comment by Kangz** [comment]
Note that WebGPU allows sampling from a mipmap while rendering to another. Chromium used to not support it but that was fixed over a year ago. @toji's [texture helper that generate mipmaps](https://github.com/toji/web-texture-tool/blob/main/src/webgpu-mipmap-generator.js) uses that feature. It would be easy to extend it to support arbitrary filters specified as a WGSL function.

IIRC sampling from a mipmap and rendering to another should also be possible in WebGL (at least with ANGLE) by specifying the `BASE_MIP_LEVEL` `texParameter`. 

> The reason is, compressed formats on the GPU are just not practical in real life. Support is not reliable enough on different platforms and the cost of having a different code path when it's supported outweighs the benefit. Also in terms of transport size, jpeg, with the right compression parameters, is still the king.

This is surprising, WebGL should have good support for compressed formats, and something like Basis should produce smaller files from JPEG AFAIK. WebGPU will require support for compressed formats as well (at least on of BC and ETC2 probably), see #144.

> One thought which is actually already in the above: having WebGPU give a generateMipmap() method, but that function's implementation is provided by the W3C in JavaScript which in turn is what browsers will expand that call to. This way, it is a common library thingy.

Doing this is complicated because the Web platform doesn't have good facilities for shipping a JS-based JS standard library. There have been multiple attempts in the past (for streams, for key-value store, etc) but none of them panned out. Having an official repo of helpers would definitely be great though.

**Comment by krogovin** [comment]
`IIRC sampling from a mipmap and rendering to another should also be possible in WebGL (at least with ANGLE) by specifying the BASE_MIP_LEVEL texParameter.`

I don't think that function does the correct thing against `textureLod()`. The spec says:

> Do a texture lookup as in texture but with explicit level-of-detail; lod specifies λbase

I am not so sure, but perhaps with `TEXTURE_MIN_LOD` and `TEXTURE_MAX_LOD` the truth is that GL native, one would be better off using `glTextureView()`, but that is not present in WebGL2 and only present as an extension in native GLES3.x (though a fair number of platforms do support it there).

**Comment by Nehon** [comment]
> I am not so sure, but perhaps with `TEXTURE_MIN_LOD` and `TEXTURE_MAX_LOD` the truth is that GL native, one would be better off using `glTextureView()`, but that is not present in WebGL2 and only present as an extension in native GLES3.x (though a fair number of platforms do support it there).

I tried the MIN/MAX LOD way and it didn't work, But let's not derail from the topic, The point is to make this crystal clear in WebGPU specs so that implementers are not left alone with their imagination.

**Comment by magcius** [comment]
It would be nice to port the AMD SPD library to WGSL: https://gpuopen.com/fidelityfx-spd/

**Comment by jdashg** [comment]
Mipmap generation via TEXTURE_BASE_LEVEL and TEXTURE_MAX_LEVEL does work in WebGL when used to preclude feedback.
MIN/MAX_LOD are for sampling/filtering, not mip level selection per se.
Here's the test: https://github.com/KhronosGroup/WebGL/blob/main/sdk/tests/conformance2/textures/misc/immutable-tex-render-feedback.html

**Comment by kvark** [comment]
@Nehon thank you for extensive feedback!

> Pre generating mipmaps and sending them over the network is not a practical, and even a realistic solution. Seeing arguments like "it's only 1/3 of the texture data" is scary, to say the least. We are targeting any kind of hardware and any kind of network bandwidth. We can't rely in the fact that all our users will have a killing fiber connection. Most of them have a slow 3g bandwidth.

Hmm. Wouldn't it be faster to transfer higher mips first, followed by lower mips? I'd consider it a win for slow-connection users, and it's only possible if you pre-generate the mips. Your website would then always display the lowest available mip level for a texture. Once a new one comes it, you'd re-create the `GPUTextureView` and the `GPUBindGroup` for this lower mip.

> Mipmap generation HAS to be possible at runtime.
> Being able to read from a mip level and write into another WITHOUT throwing a feedback loop error. 

Yes, it is possible in WebGPU, as noted in other comments.
You can see it on https://wgpu.rs/examples/?example=mipmap, for example.

> Having 1 hardware command to generate mipmaps on regular textures (as opposed to RTT textures) is REALLY handy. 

I guess we'll be keeping our ears wide open to listen if this is demanded wildly. D3D12 and Vulkan don't have built-in facilities for this, and D3D12 doesn't even have any blitting, so it's not totally clear how important the mipmap generation would be today. With proper support for compressed textures in the API, and the Basis Universal magic on the user side, it may be better to expect all of the mips coming this way.

**Comment by Nehon** [comment]
> Hmm. Wouldn't it be faster to transfer higher mips first, followed by lower mips? I'd consider it a win for slow-connection users, and it's only possible if you pre-generate the mips. Your website would then always display the lowest available mip level for a texture. Once a new one comes it, you'd re-create the `GPUTextureView` and the `GPUBindGroup` for this lower mip.

We do send low res textures first so that the time to interaction is the shortest possible. but we don't send the whole mipmap chain.

Thanks for your answer.

**Comment by magcius** [comment]
> Having 1 hardware command to generate mipmaps on regular textures (as opposed to RTT textures) is REALLY handy.

Note that GenerateMipmaps is not a hardware command on any platform. The platforms that have it: D3D11's GenerateMipmaps and GL's glGenerateMipmaps are fully implemented in the driver, either as special shaders or as other tools.

One major reason that GenerateMipmaps was removed between D3D11 and D3D12 is because explicit APIs expose a lot more of the underlying hidden state that the driver can manage. As seen above, some methods of doing downsampling require tools like OUTPUT_ATTACHMENT. Others use a compute shader, which can require different usages.

Unless the exact implementation of GenerateMipmaps is specified in the spec, with all of the bits changing, it's hard to implement such conveniences efficiently, since it's locked into an implementation. Having a standard library for this, possibly a port of the compute shader downsampler released by AMD, would be preferable to having it in the browser in my opinion.

**Comment by Nehon** [comment]
> Unless the exact implementation of GenerateMipmaps is specified in the spec, with all of the bits changing, it's hard to implement such conveniences efficiently, since it's locked into an implementation. 
Yes that's my point. that's why I think the spec should be super specific about it. Whatever implementation used.

> Having a standard library for this, possibly a port of the compute shader downsampler released by AMD, would be preferable to having it in the browser in my opinion.
I don't get why a 3rd party library would be better in every aspect we discussed, and I fail to see the benefit of it considering everybody need to generate mip maps.

**Comment by Kangz** [comment]
Adding to V1.0 milestone instead of post-V1.0 like other feature requests because this has been a very common request so we need to make a decision on whether it is in 1.0 or not and stick to it.

**Comment by davidar** [comment]
> It would be nice to port the AMD SPD library to WGSL: https://gpuopen.com/fidelityfx-spd/

I've ported a limited version of SPD to WGSL [here](https://github.com/davidar/wgpu/commit/eb3a7d87293ce811263815a131a950c6259b22cf), but there's a few issues:

- only the LDS version of SPD is compatible, the wave intrinsics version is blocked on #1459
- it's not possible to generate more than 4 mip levels in a single pass due to the limit on the number of storage texture bindings, at least until #822 is resolved
- even without that limit, SPD wouldn't be able to generate more than 6 mip levels in a single pass without #1772

**Comment by Kangz** [comment]
Thanks for the feedback! I don't think #822 would allow more storage bindings overall unfortunately, just grouping them as an array for convenience. How would read-write storage textures allow doing more than 6 mips at once? The number of storage texture bindings is a limit that can be increased if the hardware supports it too.

Do you have some performance numbers? Esp. compared to doing a repeated manual blit?

**Comment by davidar** [comment]
The first 6 mips only require workgroup shared memory, but mips 7-12 need to sample from the 6th mip texture (so need read-write in order to avoid a second dispatch). Though perhaps a storage buffer could be used as a workaround to handle the communication between workgroups.

I haven't looked at the performance yet, I suspect the port still needs some work to be competitive. I'll update with some numbers when I have a chance

**Comment by davidar** [comment]
I've updated the code to lift those last two limitations (thanks for the clarification), and done a quick performance comparison. The total cost of the draws/dispatch is similar between the two, most of the time savings are from avoiding the need to setup multiple passes. For generating 10 mip levels, the SPD port takes 0.3-0.4ms, whereas repeated blit takes about 2ms.

**Comment by Kangz** [comment]
That's quite a perf difference! Thanks for sharing. I assume that the textures you generate mipmaps for have the `STORAGE_BINDING` usage? So if generating for readonly textures that might be pessimizing future reads and a copy into an otherwise readonly texture could be needed. (that's the case for repeated blits too though)

**Comment by davidar** [comment]
> I assume that the textures you generate mipmaps for have the STORAGE_BINDING usage? So if generating for readonly textures that might be pessimizing future reads and a copy into an otherwise readonly texture could be needed. (that's the case for repeated blits too though)

Yep, the only different from repeated blitting with a fragment shader is that it's using `STORAGE_BINDING` instead of `RENDER_ATTACHMENT`

> You'll need to mark the 6th mip texture as `coherent` and insert a `memoryBarrierImage()` between the 6-level write op and 7-level read op, this might come with a penalty.

Thanks. I'm guessing `storageBarrier` would be the WGSL equivalent here? (I'm using a storage buffer rather than a texture for the communication between mips 6 and 7, due to the read-write limitation.) I think buffers are currently coherent by default (#1621)

**Comment by devshgraphicsprogramming** [comment]
> I'm using a storage buffer rather than a texture for the communication between mips 6 and 7, due to the read-write limitation.

You can't have read-write access on a storage image in WebGPU !?

**Comment by Kangz** [comment]
Unfortunately not, it isn't a required capability in Metal. We gathered some statistics on the availability in Chrome macOS installs and there's about 10% of systems with no support at all, and another 50% systems that would support only `r32float` for read-write storage.

**Comment by maierfelix** [comment]
While `glGenerateMipmap` is heavily used in WebGL, it is a legacy method and in my opinion, the implementation of it should be left to the authors of WebGPU-based libraries. `glGenerateMipmap` only works reliable in a very limited subset of cases, while in modern graphics, there are too many different and specialised scenarios to cover in a spec. It is a handy method to have during development, but the performance characteristics are too intransparent and the results are too unreliable, or not usable at all.

@kvark 
> Most 3D applications that are performance-sensitive wouldn't want to generate mipmaps at run-time, except for some cases where texture content is not known in advance (procedurally generated, captured from the environment, etc).

Outside of regular textures, there are many use cases for generating/re-generating mipmaps at runtime, often to traverse hierarchical data structures or to approximate light traversal, e.g. LPV, VXCT (or generally where a volumetric representation is used to approximate a scene for e.g. lighting).

Also as a side note, I found that up to today, manually generating mipmaps in WebGL **without** `glGenerateMipmap` is surprisingly hard. AFAIK there are still vendors who don't correctly implement the `TEXTURE_BASE_LEVEL` texture parameter, which is necessary to efficiently generate mipmaps at runtime without double buffering (and without `glGenerateMipmap`). [Here](https://jsfiddle.net/xma44/v4ud2bx9/1/) is an example to test if `TEXTURE_BASE_LEVEL` works in your browser.

**Comment by greggman** [comment]
Late to this discussion and not sure why it popped up in my sites but ...

1. I'm on the side of this belongs in the user code, not in WebGPU for all the reasons mentioned above.

2. Similar to @toji's example (which I hadn't seen) I wrote a generator. [it's here](https://github.com/greggman/webgpu-utils). The mipmap generator is a single function that generates from mip level 0 to `texture.mipLevelCount`. At the moment it also handles 2d-arrays. I thought about adding 3d support but I suspect the usage for 3d mips is low. Also, there's a bunch of image/canvas/video loading functions which make it trivial to load one or more into a texture and then generate mips.

**Comment by tsherif** [comment]
Found this discussion super interesting, as I was curious why WebGPU couldn't generate mipmaps. One suggestion I have is to consolidate the arguments for excluding mipmap generation from WebGPU in one place for easy reference, if that hasn't been done already (maybe @greggman's [WebGPU from WebGL](https://webgpufundamentals.org/webgpu/lessons/webgpu-from-webgl.html) article?). I suspect this question will come up a lot, and this thread is a long, convoluted way to get to an understanding of the situation.

**Comment by JolifantoBambla** [comment]
I also ported SPD to WGSL, mainly so I can downsample my depth buffer but it can also generate mipmaps. It's on npm and [here](https://github.com/JolifantoBambla/webgpu-spd) if anyone needs it :)

**Comment by toji** [comment]
That's amazing, thanks for sharing!

**Comment by greggman** [comment]
solution discussion thread [here](https://github.com/gpuweb/gpuweb/discussions/5253)

**Comment by AshleyScirra** [comment]
Based on that discussion, it does look to me like it's really difficult to design a general-purpose library that is consistently high-performance across hardware and supports arbitrary sizes, formats, and filtering modes. Based on results there a particular problem seems to be that render passes are faster on some devices, and compute shaders faster on other devices.

In my view it would be useful for WebGPU to provide at least something to help with this - perhaps the aforementioned image-to-image filtering blit would be a good step?

**Comment by greggman** [comment]
In my view it would first be useful to provide a library that is handles this. You can have that today. As you have experienced first hand, asking browser to fix perf issues can take years. Given you can implement a fast mipmap solution today, a solution today seems better far better than a solution several years from now over which you have no control

**Comment by devshgraphicsprogramming** [comment]
> Based on that discussion, it does look to me like it's really difficult to design a general-purpose library that is consistently high-performance across hardware and supports arbitrary sizes, formats, and filtering modes. Based on results there a particular problem seems to be that render passes are faster on some devices, and compute shaders faster on other devices.
> 
> In my view it would be useful for WebGPU to provide at least something to help with this - perhaps the aforementioned image-to-image filtering blit would be a good step?

if you're filterting nPOT textures, then you can't really do much better than a naive Blit approach most people doing Vulkan-Tutorial do.

Sure, you can improve upon it by using a shader, which gives you the ability to:
- workaround certain image formats not being renderable (RGB9E5, block compressed formats) and therefore unable to be used as a blit destination
- use a custom filtering kernel (tbh simple linear interpolation sucks)

Using a compute shader just gives you the ability to preload the filtering footprint for a block of output texels into shared memory, but you still need to do 1 mip layer per workgroup, because what SPD does is really really specific to filter kernels which are 2x2 and PoT images (so high level mip texels align neatly over low level / high res texels).

due to the above its best done as a WebGPU library and not in the WebGPU implementation, just how AMD SPD is a Vulkan library and not a magical thing sitting in the driver.
